File: src/TiltedModelGeneration/SingleRingGeneration.js
'use strict';

// =============================================================================
// SingleRingGeneration.js
// High-fidelity port of src/TiltedRingModelGeneration/SingleRingGeneration.f
// (SingleRingGenerationMod)
//
// PORTING NOTES
// -------------
// Fortran `real` → Math.fround() at each arithmetic step.
// Trig functions (sin, cos, atanh, sqrt) operate on f32 inputs and their
// outputs are wrapped in f32 to match Fortran `real` precision.
//
// RNG: Fortran threads a single `idum` integer through ran2() and gasdev().
// In JS, pass the `rng` object from makeRng() (random.js) — rng.ran2() and
// rng.gasdev() share state exactly as Fortran's idum does.
//
// KNOWN ISSUE — delZ uninitialized in Ring_CalcParticle_VSys:
// The Fortran source assigns delZ in a commented-out line:
//   c      delZ=abs(R%P(PartID)%AngPos(2))-R%zGradiantStart
// but the branch that uses delZ:
//   if(abs(...AngPos(2)) .gt. R%zGradiantStart) VRotP=VRotP-R%dvdz*delZ
// references it uninitialized. In practice dvdz=0 in the standard fit so
// this branch has no effect. Ported exactly as-is with delZ=0 as the JS
// default (matching typical Fortran stack behaviour for local reals).
//
// Pi from CommonConsts: f32(Math.PI)
// =============================================================================

const f32  = Math.fround;
const Pi   = f32(Math.PI);

const { ring_ParticleAllocation } = require('../ObjectDefinitions/TiltedRing.js');


// ---------------------------------------------------------------------------
// ring_CalcNumParticles
// Fortran: Ring_CalcNumParticles(R, cmode, CloudSurfDens)
//
// Calculates the number of particles needed to represent the ring,
// proportional to ring area and surface density.
//
// nParticles = int(CloudSurfDens * Sigma^cmode * Pi*(Rh^2 - Rl^2)) + 1
// ---------------------------------------------------------------------------
function ring_CalcNumParticles(r, cmode, cloudSurfDens) {
  const rl          = f32(f32(r.rmid) - f32(f32(r.rwidth) / f32(2.0)));
  const rh          = f32(f32(r.rmid) + f32(f32(r.rwidth) / f32(2.0)));
  const pixelRing   = f32(Pi * f32(f32(rh * rh) - f32(rl * rl)));
  const densMulti   = f32(f32(cloudSurfDens) * f32(r.sigma ** cmode));
  r.nParticles      = Math.trunc(f32(densMulti * pixelRing)) + 1;
}


// ---------------------------------------------------------------------------
// ring_ParticleGeneration
// Fortran: Ring_ParticleGeneration(R, idum)
//
// Generates all particles for a ring:
//   1. Select random position (cylindrical coords + Cartesian)
//   2. Project to sky plane (inclination + PA rotation)
//   3. Shift to galaxy center
//   4. Calculate line-of-sight velocity
//   5. Assign flux weight
//
// rng: makeRng() object from random.js (shared ran2/gasdev state)
// ---------------------------------------------------------------------------
function ring_ParticleGeneration(r, rng) {
  const rmin = f32(f32(r.rmid) - f32(f32(r.rwidth) / f32(2.0)));
  const rmax = f32(f32(r.rmid) + f32(f32(r.rwidth) / f32(2.0)));
  const area = f32(Pi * f32(f32(rmax * rmax) - f32(rmin * rmin)));

  for (let i = 0; i < r.nParticles; i++) {
    ring_ParticlePosSelect(r, rng, rmin, rmax, i);
    particlePosProject(r.p[i], r.inclination, r.positionAngle);
    particlePos_NewCenter(r.p[i], r.centPos);
    ring_CalcParticle_VSys(r, i, rng);
    ring_CalcParticleFlux_Basic(r, i, area);
  }
}


// ---------------------------------------------------------------------------
// ring_ParticlePosSelect
// Fortran: Ring_ParticlePosSelect(R, idum, Rmin, Rmax, PartID)
//
// Selects a random position for particle PartID:
//   RR    = sqrt(ran2*(Rmax^2 - Rmin^2) + Rmin^2)  equal-area radius
//   Theta = ran2 * 2*Pi                              uniform angle
//   Z     = atanh(2*ran2 - 1) * z0                  sech^2 height
//
// Stores cylindrical (RR, Theta, Z) in angPos and Cartesian in pos.
// ---------------------------------------------------------------------------
function ring_ParticlePosSelect(r, rng, rmin, rmax, partID) {
  const p = r.p[partID];

  // Equal-area radius sampling
  const rr = f32(Math.sqrt(f32(
    f32(rng.ran2() * f32(f32(rmax * rmax) - f32(rmin * rmin)))
    + f32(rmin * rmin)
  )));

  // Uniform angle
  const theta = f32(f32(rng.ran2()) * f32(2.0 * Pi));

  // sech^2 height via atanh
  const z = f32(f32(Math.atanh(f32(f32(2.0) * f32(rng.ran2()) - f32(1.0))))
    * f32(r.z0));

  // Store cylindrical coordinates
  p.angPos[0] = rr;
  p.angPos[1] = theta;
  p.angPos[2] = z;

  // Store Cartesian coordinates
  p.pos[0] = f32(rr * f32(Math.cos(theta)));
  p.pos[1] = f32(rr * f32(Math.sin(theta)));
  p.pos[2] = z;
}


// ---------------------------------------------------------------------------
// particlePosProject
// Fortran: ParticlePosProject(P, Inclination, PositionAngle)
//
// Projects particle from ring plane to sky plane:
//   1. Incline: XTemp=Pos[0], YTemp=Pos[1]*cos(i) - Pos[2]*sin(i)
//   2. Rotate by PA:
//      ProjectedPos[0] = XTemp*cos(PA) - YTemp*sin(PA)
//      ProjectedPos[1] = XTemp*sin(PA) + YTemp*cos(PA)
// ---------------------------------------------------------------------------
function particlePosProject(p, inclination, positionAngle) {
  const xTemp = f32(p.pos[0]);
  const yTemp = f32(
    f32(f32(p.pos[1]) * f32(Math.cos(inclination)))
    - f32(f32(p.pos[2]) * f32(Math.sin(inclination)))
  );

  const cpa = f32(Math.cos(positionAngle));
  const spa = f32(Math.sin(positionAngle));

  p.projectedPos[0] = f32(f32(xTemp * cpa) - f32(yTemp * spa));
  p.projectedPos[1] = f32(f32(xTemp * spa) + f32(yTemp * cpa));
}


// ---------------------------------------------------------------------------
// particlePos_NewCenter
// Fortran: ParticlePos_NewCenter(P, NewCent)
//
// Shifts projected position by galaxy center offset.
// ---------------------------------------------------------------------------
function particlePos_NewCenter(p, newCent) {
  p.projectedPos[0] = f32(f32(p.projectedPos[0]) + f32(newCent[0]));
  p.projectedPos[1] = f32(f32(p.projectedPos[1]) + f32(newCent[1]));
}


// ---------------------------------------------------------------------------
// ring_CalcParticle_VSys
// Fortran: Ring_CalcParticle_VSys(R, PartID, idum)
//
// Calculates line-of-sight velocity for particle PartID:
//   vLOS = VSys
//        + VRot * cos(theta) * sin(incl)     (rotation)
//        + VRad * sin(theta) * sin(incl)     (radial)
//        + Vvert * cos(incl)                 (vertical bulk motion)
//        + gasdev() * VDisp                  (dispersion scatter)
//
// NOTE: The dvdz vertical gradient branch references `delZ` which is
// uninitialized in the Fortran source (assignment line is commented out).
// In the standard fit dvdz=0, so the branch is never entered. Ported
// with delZ=0 matching typical Fortran local variable initialisation.
// ---------------------------------------------------------------------------
function ring_CalcParticle_VSys(r, partID, rng) {
  const p = r.p[partID];

  let vRotP  = f32(r.vRot);
  const vRadP  = f32(r.vRad);
  const vDispP = f32(r.vDisp);

  // Vertical velocity gradient (dvdz branch)
  // delZ is uninitialized in Fortran — use 0 matching standard fit behaviour
  const delZ = f32(0.0);
  if (f32(Math.abs(p.angPos[2])) > f32(r.zGradiantStart)) {
    vRotP = f32(vRotP - f32(f32(r.dvdz) * delZ));
  }

  const cTheta = f32(Math.cos(p.angPos[1]));
  const sTheta = f32(Math.sin(p.angPos[1]));

  const vFromRotation = f32(f32(vRotP * cTheta) * f32(Math.sin(r.inclination)));
  const vFromRadial   = f32(f32(vRadP * sTheta) * f32(Math.sin(r.inclination)));
  const vFromVertical = f32(f32(r.vvert) * f32(Math.cos(r.inclination)));

  p.projectedVel[2] = f32(
    f32(f32(r.vSys) + vFromRotation)
    + f32(vFromRadial + vFromVertical)
  );

  // Gaussian dispersion scatter
  p.projectedVel[2] = f32(
    f32(p.projectedVel[2]) + f32(f32(rng.gasdev()) * vDispP)
  );
}


// ---------------------------------------------------------------------------
// ring_CalcParticleFlux_Basic
// Fortran: Ring_CalcParticleFlux_Basic(R, PartID, Area)
//
// Assigns flux to particle — equal share of total ring flux.
// Flux = Sigma * Area / nParticles
// ---------------------------------------------------------------------------
function ring_CalcParticleFlux_Basic(r, partID, area) {
  const sigmaP = f32(r.sigma);
  r.p[partID].flux = f32(f32(sigmaP * area) / f32(r.nParticles));
}


// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------
module.exports = {
  ring_CalcNumParticles,
  ring_ParticleGeneration,
  ring_ParticlePosSelect,
  particlePosProject,
  particlePos_NewCenter,
  ring_CalcParticle_VSys,
  ring_CalcParticleFlux_Basic
};


// ---------------------------------------------------------------------------
// Self-test (node SingleRingGeneration.js)
// ---------------------------------------------------------------------------
if (require.main === module) {
  const { makeRng }             = require('../StandardMath/random.js');
  const { Ring }                = require('../ObjectDefinitions/TiltedRing.js');
  const { ring_ParticleAllocation } = require('../ObjectDefinitions/TiltedRing.js');

  // Build a simple test ring
  const r           = new Ring();
  r.rmid            = f32(10.0);
  r.rwidth          = f32(2.0);
  r.sigma           = f32(0.01);
  r.inclination     = f32(45.0 * Math.PI / 180.0);
  r.positionAngle   = f32(30.0 * Math.PI / 180.0);
  r.centPos[0]      = f32(50.0);
  r.centPos[1]      = f32(50.0);
  r.vSys            = f32(1000.0);
  r.vRot            = f32(150.0);
  r.vRad            = f32(0.0);
  r.vDisp           = f32(8.0);
  r.vvert           = f32(0.0);
  r.dvdz            = f32(0.0);
  r.z0              = f32(0.0);
  r.zGradiantStart  = f32(0.0);

  console.log('=== ring_CalcNumParticles ===');
  ring_CalcNumParticles(r, 1, f32(1.0));
  console.log('nParticles:', r.nParticles);

  console.log('\n=== ring_ParticleGeneration (seed -1) ===');
  const rng = makeRng(-1);
  ring_ParticleAllocation(r);
  ring_ParticleGeneration(r, rng);

  console.log('particle[0]:');
  console.log('  angPos:       ', Array.from(r.p[0].angPos).map(v => v.toFixed(6)));
  console.log('  projectedPos: ', Array.from(r.p[0].projectedPos).map(v => v.toFixed(6)));
  console.log('  projectedVel[2]:', r.p[0].projectedVel[2].toFixed(4));
  console.log('  flux:         ', r.p[0].flux.toExponential(6));

  // Flux conservation check
  let totalFlux = f32(0.0);
  for (let i = 0; i < r.nParticles; i++) totalFlux = f32(totalFlux + r.p[i].flux);
  const area = f32(Pi * f32(f32(f32(r.rmid + r.rwidth/2)**2) - f32(f32(r.rmid - r.rwidth/2)**2)));
  const expectedFlux = f32(r.sigma * area);
  console.log('\n=== flux conservation ===');
  console.log('total flux:   ', totalFlux.toExponential(6));
  console.log('sigma*area:   ', expectedFlux.toExponential(6));
  console.log('match:', Math.abs(totalFlux - expectedFlux) < 1e-5 ? 'OK' : 'FAIL');
}

File: src/TiltedRingModelGeneration/SingleRingGeneration.f
cccccccccccccccccccccccccccccccccccccccccccccc
c
c     This module contains the routines needed to
c       fill a single ring with the particles necessary for
c       a tilted ring model
c
ccccccccccccccccccccccccccccccccccccccccccccccccc

      module SingleRingGenerationMod
      use ParticleMod
      use TiltedRingMod


      implicit none

      contains

ccccccc
c       RoundForParticleStability: masks off the low 12 bits of x's
c           float32 mantissa (keeping the top 11 of 23 bits -- ~0.05%
c           relative precision) before Ring_CalcNumParticles truncates it
c           to an integer particle count.
c
c           Root cause this exists for: R%Sigma (and everything derived
c           from it, e.g. DensMultiplications below) can differ between
c           Fortran and the JS port by a handful of float32 ULPs (~5e-7
c           relative) -- an already-known, unavoidable cross-platform
c           floating-point difference (compiler codegen, not a bug --
c           see UPSTREAM_SYNC.md) that is normally harmless everywhere
c           else in this pipeline, since it stays a smoothly tiny
c           difference in whatever value carries it. But
c           Ring_CalcNumParticles's int(...)+1 is a hard truncation, and a
c           continuous value that happens to land within noise-distance of
c           an integer boundary truncates to *different whole integers* on
c           the two platforms. That single-particle difference then
c           consumes a different number of ran2()/gasdev() draws,
c           permanently desyncing idum for the rest of the fit --
c           confirmed directly: Fortran and the JS port ran in bit-exact
c           lockstep (identical idum every evaluation) until this exact
c           failure mode first triggered, then diverged permanently and
c           the JS optimizer never recovered.
c
c           Masking gives ~900x safety margin over the observed ~5e-7
c           relative noise, at the cost of ~0.05% precision on a Monte
c           Carlo sampling-density parameter -- scientifically
c           irrelevant, and it must be applied identically on both
c           platforms (see hexF32/roundForParticleStability in
c           SingleRingGeneration.js) for the masked value itself to still
c           agree.
      real function RoundForParticleStability(x)
      implicit none
      real, INTENT(IN) :: x
      integer(4) ix
      integer(4), parameter :: MASK = int(z'FFFFF000',4)
      ix = transfer(x, 0)
      ix = iand(ix, MASK)
      RoundForParticleStability = transfer(ix, 0.0)
      return
      end function
cccccccc

ccccccc
      subroutine Ring_CalcNumParticles(R,cmode,CloudSurfDens,Noise
     &          ,AvgChannelsPerPix)
      use CommonConsts
      use PipelineGlobals
      implicit none
      Type(Ring), INTENT(INOUT):: R
      integer,INTENT(IN) ::cmode
      real, INTENT(IN) :: CloudSurfDens
      real,INTENT(IN) :: Noise
      real,INTENT(IN) :: AvgChannelsPerPix
      real Pixel_Ring
      real DensMultiplications
      real Rl,Rh

c           Get the rough area of the ring in pixels
c      Pixel_Ring=2.*Pi*R%Rmid*R%Rwidth /pixelarea      !Original in 3DBarolo GalMod
      Rl=R%Rmid-R%Rwidth/2.
      Rh=R%Rmid+R%Rwidth/2.
c      Pixel_Ring=Pi*(Rh**2.-Rl**2)/pixelarea
      Pixel_Ring=Pi*(Rh**2.-Rl**2)
c       Calculate a term to get the rough number of clouds per pixel area.  It is normalized by
c           the surface density so that each particle has roughly the
c           same amount of flux.
c
c           DELIBERATELY on the pre-upstream-commit-76ade48 formula for
c           now (not using Noise/AvgChannelsPerPix below) -- Nathan's own
c           attached example config used cdens=10, but his email text
c           says the actual default is cdens=100 (10 looks like a typo in
c           the example, not the intended default); revisit once that's
c           confirmed and re-verify Fortran/JS agreement before switching
c           this back on. Noise/AvgChannelsPerPix stay threaded through
c           Ring_CalcNumParticles's signature (unused here) so re-enabling
c           is a one-line change, not a re-plumb -- see
c           TiltedRingModelGeneration.f's CalcAvgChanPerPix, already
c           wired and confirmed bit-matching between Fortran and the JS
c           port.
      DensMultiplications=CloudSurfDens
     &               *((R%Sigma)**real(cmode))
      R%nParticles=int(RoundForParticleStability(DensMultiplications
     &          *Pixel_Ring))+1
      if (TraceSwitch .eq. 1) then
        print*, "NPTRACE",R%Rmid,R%Sigma,DensMultiplications,
     &          Pixel_Ring,AvgChannelsPerPix,R%nParticles
        print '(A,Z8.8,1X,Z8.8,1X,Z8.8,1X,Z8.8,1X,Z8.8)',
     &      "NPHEX",transfer(R%Sigma,0),transfer(Noise,0),
     &      transfer(DensMultiplications,0),
     &      transfer(Pixel_Ring,0),transfer(AvgChannelsPerPix,0)
      endif
c      print*, "Single Ring PA",R%PositionAngle*180./Pi
c      print*, "Single Ring Center",R%CentPos
c      print*, "Number of particles", R%nParticles,R%Sigma
c      print*, R%Rwidth,R%Rmid,R%Sigma,CloudSurfDens,cmode
c     &          ,Pixel_Ring,DensMultiplications, R%PositionAngle*180./Pi

      return
      end subroutine
cccccccc

ccccccc
      subroutine Ring_ParticleGeneration(R,idum)
      use BasicRanNumGen
      use CommonConsts
      use PipelineGlobals, only: TraceSwitch
      implicit none
      Type(Ring), INTENT(INOUT) :: R     !The ring that we are generating particles for
      integer, INTENT(INOUT) :: idum          !idum is a seed for the random number generator
      integer i
      real Rmin,Rmax, Area       !The minimum and maximum radii of the ring and the ring area
      character(64) BisectEnvVal
      integer BisectEnvLen

      call get_environment_variable("TRACE_OVERRIDE_IDUM",
     &          BisectEnvVal,BisectEnvLen)

c           To randomly sample the ring area we need Rmin and Rmax
      Rmin=R%Rmid-R%Rwidth/2.       !We need Rmin and Rmax for the ring first
      Rmax=R%Rmid+R%Rwidth/2.
      Area=Pi*(Rmax**2.-Rmin**2)
c      print*, "Single Ring Area Check", Area,Rmin,Rmax
c           One-off diagnostic ("bisection paradox" investigation, Dan
c               2026-08-18): BINTRACE showed every particle landing in the
c               same cell on both platforms but with a different constant
c               Flux value for ring 3 -- since Flux=Sigma*Area/nParticles
c               and RINGTRACE only ever dumps ring 0's fields, this prints
c               every ring's own Rmid/Rwidth/Sigma/nParticles/Area so the
c               actual numeric input responsible can be identified
c               directly, gated on TRACE_OVERRIDE_IDUM like the others.
      if (BisectEnvLen .gt. 0) then
        print '(A,ES25.17,1X,ES25.17,1X,ES25.17,1X,I12,1X,ES25.17)',
     &      'RINGGEOMTRACE', R%Rmid, R%Rwidth, R%Sigma,
     &      R%nParticles, Area
      endif


c      print*, "Ring Params", R%CentPos,R%Inclination, R%PositionAngle
c     &          ,R%VSys, R%VRot, R%VRad, R%VDisp,R%Vvert
c     &          ,R%dvdz,R%Sigma,R%z0,R%zGradiantStart
c      print*, "Consistency Check", R%VSys,R%VRad
c           Now go through all the particles
      do i=0, R%nParticles-1
        call Ring_ParticlePosSelect(R,idum, Rmin,Rmax,i)
        call ParticlePosProject(R%P(i), R%Inclination, R%PositionAngle)
        call ParticlePos_NewCenter(R%P(i), R%CentPos)
        call Ring_CalcParticle_VSys(R,i,idum)
        call Ring_CalcParticleFlux_Basic(R,i, Area)
        if (TraceSwitch.eq.1 .and. i.lt.5) then
          print*, "PARTTRACE",R%Rmid,i,idum,R%P(i)%Pos(0),
     &            R%P(i)%Pos(1),R%P(i)%Pos(2),
     &            R%P(i)%ProjectedPos(0),R%P(i)%ProjectedPos(1),
     &            R%P(i)%ProjectedVel(2)
        endif
c           One-off diagnostic (Fortran-vs-JS gasdev-desync bisection,
c               Dan 2026-08-18): checkpoint every 200th particle (plus the
c               very last one) across the FULL ring, not just the first 5
c               PARTTRACE covers -- to localize exactly where a desync
c               (re)starts, if one does, deeper into a ring's particle
c               loop. Gated on TRACE_OVERRIDE_IDUM.
        if (BisectEnvLen .gt. 0 .and.
     &          (mod(i,200).eq.0 .or. i.eq.R%nParticles-1)) then
          print '(A,F10.6,1X,I6,1X,I12,1X,F14.8)', 'BISECTTRACE',
     &        R%Rmid,i,idum,R%P(i)%ProjectedVel(2)
        endif

c        print*, "Random Check", R%P(i)%AngPos, R%P(i)%Pos
c        print*, "Projected Random Check", R%P(i)%ProjectedPos(0:1)
c        print*, "PRojected Velocity", R%P(i)%ProjectedVel(2)
      enddo

c      print*, "particle Flux ChecK", sum(R%P(0:R%nParticles-1)%Flux)


      return
      end subroutine
cccccccc


ccccccc
      subroutine Ring_ParticlePosSelect(R,idum, Rmin, Rmax, PartID )
      use BasicRanNumGen
      use CommonConsts
      implicit none
      Type(Ring), INTENT(INOUT) :: R     !The ring that we are generating particles for
      integer, INTENT(INOUT) :: idum          !idum is a seed for the random number generator
      integer, INTENT(IN) :: PartID
      real, INTENT(IN) :: Rmin,Rmax
      real RR, Theta, Z     !Temporary cylindrical coordinates of each particle
      real fd_sin, fd_cos, fd_atanh
      external fd_sin, fd_cos, fd_atanh

c           First get a random radius -- however this is not uniform in R.  We want
c           equal area sampling we'll be using a sqrt distribution.
      RR=ran2(idum)*(Rmax**2.-Rmin**2.)             !Sample a uniform squared radius in the correct range
      RR=sqrt(RR+Rmin**2.)          !Add on the minimum radius squared and take the sqrt
c           Next get a random angle
      Theta=ran2(idum)*2.*Pi
c           Finally get a height from a sech^2 distribution (use atanh on a random distrubtion and scale
c                   by the height
      Z=fd_atanh((2.*ran2(idum)-1.))*R%z0
c      print*, "Z", Z,R%z0

c           Store the cylindrical coordinates
      R%P(PartID)%AngPos(0)=RR
      R%P(PartID)%AngPos(1)=Theta
      R%P(PartID)%AngPos(2)= Z
c           Store the Cartesian Coordinates
      R%P(PartID)%Pos(0)=RR*fd_cos(Theta)
      R%P(PartID)%Pos(1)=RR*fd_sin(Theta)
      R%P(PartID)%Pos(2)= Z

      return
      end subroutine
ccccccccc

cccccccc
      subroutine ParticlePosProject(P,Inclination,PositionAngle)
      implicit none
      Type(Particle), INTENT(INOUT) :: P
      real, INTENT(IN) :: Inclination, PositionAngle
      real cpa,spa,XTemp,YTemp
      real fd_sin, fd_cos
      external fd_sin, fd_cos
c       First get the inclined positions in the major axis frame (aligned with X-axis)
c           in the temporary arrays
      XTemp=P%Pos(0)
      YTemp=P%Pos(1)*fd_cos(Inclination)-P%Pos(2)*fd_sin(Inclination)
c       Now rotate the positions by the position angle
      cpa=fd_cos(PositionAngle)        !Get the cos and sin values
      spa=fd_sin(PositionAngle)
c       Calculate the coordinates in the projected array
      P%ProjectedPos(0)=XTemp*cpa-YTemp*spa
      P%ProjectedPos(1)=XTemp*spa+YTemp*cpa

c      print*, P%Pos(0:1), P%AngPos(0:2)
c     &          , P%ProjectedPos(0:1)
c      print*, P%ProjectedPos, PositionAngle*180./3.14, XTemp,YTemp
      return
      end subroutine
cccccccccccc

cccccccc
      subroutine ParticlePos_NewCenter(P,NewCent)
      implicit none
      Type(Particle), INTENT(INOUT) :: P
      real, INTENT(IN) :: NewCent(0:1)

c       Shift the projected position to lie about the new center
c      print*, "Initial Pos", P%ProjectedPos,NewCent
      P%ProjectedPos(0)=P%ProjectedPos(0)+NewCent(0)
      P%ProjectedPos(1)=P%ProjectedPos(1)+NewCent(1)
c      print*, "Final Pos", P%ProjectedPos
      return
      end subroutine
cccccccccccc

cccccc
      subroutine Ring_CalcParticle_VSys(R,PartID,idum)
      use BasicRanNumGen
      implicit none
      Type(Ring), INTENT(INOUT) :: R
      integer, INTENT(IN) :: PartID
      integer, INTENT(INOUT) :: idum          !idum is a seed for the random number generator
      real VRotP,VRadP,VVertP,VDispP
      real V_FromRotation, V_FromRadial,V_FromVertical
      real delR,delZ
      real cTheta,sTheta
      integer DerivativeSide
      real fd_sin, fd_cos
      external fd_sin, fd_cos

c           First get the particle's rotational, radial, and vertical motions
c           at the current radial positions

c               To do the linear interpolation, figure out which side you're on
c      delR=(R%P(PartID)%AngPos(0)-R%Rmid)
c      if(delR .le. 0) then
c        DerivativeSide=0
c      else
c        DerivativeSide=1
c      endif

c      delZ=abs(R%P(PartID)%AngPos(2))-R%zGradiantStart
      VRotP=R%VRot !+ R%dVRot_dR(DerivativeSide)*delR
      VRadP=R%VRad !+ R%dVRad_dR(DerivativeSide)*delR
      VDispP=R%VDisp !+ R%dVDisp_dR(DerivativeSide)*delR
c           Check if the particle is far enough above the plane that V_rot is affected
      if(abs(R%P(PartID)%AngPos(2)) .gt. R%zGradiantStart) then
        VRotP=VRotP-R%dvdz*delZ
      endif
c
c           Now get the sin and cos of the angle
c
      cTheta=fd_cos(R%P(PartID)%AngPos(1))
      sTheta=fd_sin(R%P(PartID)%AngPos(1))
c
c           Next get the observed velocity due to the main components
c
      V_FromRotation=VRotP*cTheta*fd_sin(R%Inclination)
      V_FromRadial=VRadP*sTheta*fd_sin(R%Inclination)
      V_FromVertical=R%Vvert*fd_cos(R%Inclination)         !Note the Vvert is the global vertical motion of a ring
c        Sum up the various components
      R%P(PartID)%ProjectedVel(2)=R%VSys+V_FromRotation
     &                      +V_FromRadial+V_FromVertical
c
c           Finally get a random velocity from the dispersion and add it in to the projected velocity
c
      R%P(PartID)%ProjectedVel(2)=R%P(PartID)%ProjectedVel(2)
     &                      +gasdev(idum)*VDispP

c      print*, "hmmm", VDispP, R%VDisp

      return
      end subroutine
ccccccccc

cccccc
      subroutine Ring_CalcParticleFlux_Basic(R,PartID,Area)
      implicit none
      Type(Ring), INTENT(INOUT) :: R
      integer, INTENT(IN) :: PartID
      real,INTENT(IN):: Area
      real SigmaP, del R
      integer DerivativeSide
c           Get the local Sigma
      SigmaP=R%Sigma

      R%P(PartID)%Flux=SigmaP*Area/real(R%nParticles)


      return
      end subroutine
cccccccc

      end module
